lib/pages/list/list.dart

```diff
@@ -145,60 +581,148 @@ class _ListPageState extends State<ListPage> {
-          labelText: 'Task Name',
+          labelText: (bucket != null ? '${bucket.title}: ' : '') + 'New Task Name',
           hintText: 'eg. Milk',
         ),
       ),
     );
   }
 
-  _addItem(String title, BuildContext context) {
+  Future<void> _addItem(String title, BuildContext context, [Bucket bucket]) {
     var globalState = VikunjaGlobal.of(context);
     var newTask = Task(
       id: null,
       title: title,
       createdBy: globalState.currentUser,
       done: false,
+      bucketId: bucket?.id,
     );
     setState(() => _loadingTasks.add(newTask));
-    Provider.of<ListProvider>(context, listen: false)
+    return Provider.of<ListProvider>(context, listen: false)
         .addTask(
       context: context,
       newTask: newTask,
       listId: _list.id,
     )
         .then((_) {
       ScaffoldMessenger.of(context).showSnackBar(SnackBar(
-        content: Text('The task was added successfully!'),
+        content: Text('The task was added successfully' + (bucket != null ? ' to ${bucket.title}' : '') + '!'),
       ));
       setState(() {
         _loadingTasks.remove(newTask);
       });
     });
   }
+
+  Future<void> _addBucketDialog(BuildContext context) {
+    FocusScope.of(context).unfocus();
+    return showDialog(
+      context: context,
+      builder: (_) => AddDialog(
+        onAdd: (title) => _addBucket(title, context),
+        decoration: InputDecoration(
+          labelText: 'New Bucket Name',
+          hintText: 'eg. To Do',
+        ),
+      )
+    );
+  }
+
+  Future<void> _addBucket(String title, BuildContext context) {
+    return Provider.of<ListProvider>(context, listen: false).addBucket(
+      context: context,
+      newBucket: Bucket(
+        id: null,
+        title: title,
+        createdBy: VikunjaGlobal.of(context).currentUser,
+        listId: _list.id,
+      ),
+      listId: _list.id,
+    ).then((_) {
+      ScaffoldMessenger.of(context).showSnackBar(SnackBar(
+        content: Text('The bucket was added successfully!'),
+      ));
+      setState(() {});
+    });
+  }
+
+  Future<void> _updateBucket(BuildContext context, Bucket bucket) {
+    return Provider.of<ListProvider>(context, listen: false).updateBucket(
+      context: context,
+      bucket: bucket,
+    ).then((_) {
+      ScaffoldMessenger.of(context).showSnackBar(SnackBar(
+        content: Text('${bucket.title} bucket updated successfully!'),
+      ));
+      setState(() {});
+    });
+  }
+
+  Future<void> _deleteBucket(BuildContext context, Bucket bucket) async {
+    await Provider.of<ListProvider>(context, listen: false).deleteBucket(
+      context: context,
+      listId: bucket.listId,
+      bucketId: bucket.id,
+    ).then((_) {
+      ScaffoldMessenger.of(context).showSnackBar(SnackBar(
+        content: Row(
+          children: <Widget>[
+            Text('${bucket.title} was deleted.'),
+            Icon(Icons.delete),
+          ],
+        ),
+      ));
+    });
+    _onViewTapped(1);
+  }
 }
```

Comment:
Is it possible to move all of the bucket helper functions into their own file? There isn't a reason for such large files, they just get less readable

Reply:
I think to do it properly would require getting state management sorted out in conjunction. All of this was just to get everything to a working point. Also the project's directory/file structure doesn't lend itself to widgets being broken up easily. What I want to do is break out the list and bucket views from the page and have them handle their own state where possible.

Reply:
That's fair enough, I think the code structure needs a big overhaul in general.